### PR TITLE
Send a stable style object reference to wrapped components

### PR DESCRIPTION
**Pull request by jamesism**
# Summary

This PR modifies `useRestyle` in order to:
* add a step to `fitlerRestyleProps` that will build a string out of the restyle specific props.
* use the serialized string as the dependency to the `useMemo` to recalculate the style object

This means that only when: 
* the value of restyle props changes
* the theme changes
* the dimensions change, or 
* the `style` prop passed to the wrapper changes

will we call `buildStyle` and generate a new reference for the `style` prop that ultimately gets passed to the wrapped component.

This allows our wrapped components to memoize effectively and avoid unnecessary renders:

https://user-images.githubusercontent.com/887305/155625270-3bd076ef-7f97-4ef4-8110-8b54519f1eeb.mov

https://user-images.githubusercontent.com/887305/155625930-fdd7382f-ca9b-4f16-86f7-b2076a49f217.mov

The first video shows `Button` appearing on WYDR plugin as rendering unnecessarily primarily due to the `style` object changing. The second shows the library delivering a stable reference and using `memo` on the wrapped `Button` component, which eliminates the unnecessary renders from WDYR.

### CODE REVIEW

File: src/hooks/useRestyle.ts [hunk 1]
```diff
@@ -51,24 +60,30 @@ const useRestyle = <
...
+    const style = composedRestyleFunction.buildStyle(restylePropsRef.current!, {
       theme,
       dimensions,
     });
 
-    cleanProps.style = [style, props.style].filter(Boolean);
-    return cleanProps;
-  }, [composedRestyleFunction, props, dimensions, theme]);
+    return [style, props.style].filter(Boolean);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (sbalay) [inline src/hooks/useRestyle.ts:80]: Might be worth a code comment explaining why we disable this rule here.

**Review by sbalay — APPROVED** [review]
Great job!! 🚀 🚀 🚀 

Made some minor suggestions

File: src/hooks/useRestyle.ts [hunk 2]
```diff
@@ -51,24 +60,30 @@ const useRestyle = <
...
+  restylePropsRef.current = restyleProps as TProps;
 
-  const restyled = useMemo(() => {
-    const {cleanProps, restyleProps} = filterRestyleProps(
-      props,
-      composedRestyleFunction.propertiesMap,
-    );
-    const style = composedRestyleFunction.buildStyle(restyleProps, {
+  const calculatedStyle = useMemo(() => {
+    const style = composedRestyleFunction.buildStyle(restylePropsRef.current!, {
```
[the remark this answers is not in the document]
Comment (jamesism) [reply src/hooks/useRestyle.ts:74]: You're right, this ref is not necessary and I'll remove :+1: started with a more ref based approach 😅

**Comment by jamesism** [comment]
This stabilizes `style` considerably, but I think there could still be a gap if `dimensions` or `theme` were to have stable values but an unstable reference. To mitigate this another approach to consider might be to do away with the `useMemo` entirely, and instead stabilize the result of `buildStyle` with the same string serialization technique and a stored object reference.

**Comment by sbalay** [comment]
> This stabilizes `style` considerably, but I think there could still be a gap if `dimensions` or `theme` were to have stable values but an unstable reference. If this is a problem another approach to consider might be to do away with the `useMemo` entirely, and instead stabilize the result of `buildStyle` with the same string serialization technique and a stored object reference.

But we do have control over dimensions and theme to make sure they're stable references, right?

Dimensions seems to be ok and theme being stable or not depends on dev using the ThemeProvider with a static object reference

**Comment by jamesism** [comment]
> Dimensions seems to be ok and theme being stable or not depends on dev using the ThemeProvider with a static object reference

Yup, you're right, this is likely negligible :+1:

**Comment by linddominic** [comment]
Can we measure if this adds any time to render?

**Review by linddominic — APPROVED** [review]

**Comment by Braden1996** [comment]
@jamesism, sorry for the comment and tag here. Just wondering where you found that WDYR Flipper plugin? I can't seem to find it online; is it something you have internally?

**Pull request merged**